### PR TITLE
Refactor/validation framework

### PR DESCRIPTION
**Pull request by btomy**

### CODE REVIEW

File: packages/cads-validation/lib/component.validation.scss [hunk 1]
```diff
@@ -7,21 +7,28 @@
...
+    display: none;
 
-  //need to add the warning icon in here once the icons have been pushed
-  // &::before{
-  //     content:url('/assets/images/warning.svg');
-  //     margin-right: 5px;
-  // }
-}
\ No newline at end of file
+    &.active {
```
Comment (ChrisBrettCA) [inline packages/cads-validation/lib/component.validation.scss]: @btomy the `error` class is spot on, let's do a similar thing with the `active` class too.
```
&--active { ... }
```
Obvs need a markup change too. :)

**Comment by ChrisBrettCA** [comment]
In order to view these components add this to sandbox `HTML`:

```
<!-- you're directly inside the <body /> here... -->
<div class="c-wrap">
  <div class="c-wrap">
    <div class="o-grid">
      <div class="o-grid__unit o-grid__unit--t-1-2">

        <!-- RADIO BUTTONS AND CHECKBOXES -->
        <div class="h2">Radio/Checkboxes (default)</div>
        <fieldset class="c-fieldset">
          <legend class="c-label">Did this help?</legend>

          <div class="c-fieldset__item">
            <div class="c-radio c-radio--invalid">
              <input checked id="yes" value="yes" name="feedback" type="radio">
              <label for="yes">Yes</label>
            </div>
          </div>

          <div class="c-fieldset__item">
            <div class="c-radio c-radio--invalid">
              <input id="no" value="no" name="feedback" type="radio">
              <label for="no">No</label>
            </div>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select an option</div>
        </fieldset>

        <fieldset class="c-fieldset">
          <legend class="c-fieldset__label">Right and/or left?</legend>

          <div class="c-fieldset__item">
            <div class="c-checkbox">
              <input checked id="right" value="right" name="direction" type="checkbox">
              <label for="right">Right</label>
            </div>
          </div>

          <div class="c-fieldset__item">
            <div class="c-checkbox c-checkbox--invalid">
              <input id="left" value="left" name="direction" type="checkbox">
              <label for="left">Left</label>
            </div>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select an option</div>
        </fieldset>

        <!-- INPUT FIELDS -->
        <div class="h2">Input fields (default)</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="postcode">What's your postcode?</label>
          <span class="c-label__metainfo">For example: SW1A 2AA</span>
          <div class="c-input c-input--invalid">
            <input id="postcode" type="text" name="postcode" pattern="[a-z]{1,15}" />
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please enter a valid postcode</div>
        </fieldset>

        <!-- SELECT DROPDOWNS -->
        <div class="h2">Select menus</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="select1">Example dropdown select</label>
          <div class="c-select c-select--invalid">
            <select id="select1">
              <option value="Please select">Please select</option>
              <option value="Option #1">Option #1</option>
              <option value="Option #2">Option #2</option>
              <option value="Option #3">Option #3</option>
              <option value="Option #4">Option #4</option>
              <option value="Option #5">Option #5</option>
            </select>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select one option</div>
        </fieldset>

        <!-- TEXTAREAS -->
        <div class="h2">Textarea (default)</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="message">Message:</label>
          <div class="c-textarea c-textarea--invalid">
            <textarea id="message"></textarea>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please leave a message</div>
        </fieldset>
      </div>

      <div class="o-grid__unit o-grid__unit--t-1-2">

        <!-- RADIO BUTTONS AND CHECKBOXES (SM) -->
        <div class="h2">Radio/Checkboxes (small)</div>
        <fieldset class="c-fieldset">
          <legend class="c-label">Did this help?</legend>

          <div class="c-fieldset__item">
            <div class="c-radio c-radio--sm c-radio--invalid">
              <input checked id="yesSm" value="yes" name="feedback" type="radio">
              <label for="yesSm">Yes</label>
            </div>
          </div>

          <div class="c-fieldset__item">
            <div class="c-radio c-radio--sm c-radio--invalid">
              <input id="noSm" value="no" name="feedback" type="radio">
              <label for="noSm">No</label>
            </div>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select an option</div>
        </fieldset>

        <fieldset class="c-fieldset">
          <legend class="c-fieldset__label">Right and/or left?</legend>

          <div class="c-fieldset__item">
            <div class="c-checkbox c-checkbox--sm">
              <input checked id="rightSm" value="right" name="direction" type="checkbox">
              <label for="rightSm">Right</label>
            </div>
          </div>

          <div class="c-fieldset__item">
            <div class="c-checkbox c-checkbox--sm c-checkbox--invalid">
              <input id="leftSm" value="left" name="direction" type="checkbox">
              <label for="leftSm">Left</label>
            </div>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select an option</div>
        </fieldset>

        <!-- INPUT FIELDS (SM) -->
        <div class="h2">Input fields (small)</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="postcodeSm">What's your postcode?</label>
          <span class="c-label__metainfo">For example: SW1A 2AA</span>
          <div class="c-input c-input--sm c-input--invalid">
            <input id="postcodeSm" type="text" name="postcode" pattern="[a-z]{1,15}" />
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please enter a valid postcode</div>
        </fieldset>

        <!-- SELECT DROPDOWNS (SM) -->
        <div class="h2">Select menus (small)</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="select1Sm">Example dropdown select</label>
          <div class="c-select c-select--sm c-select--invalid">
            <select id="select1Sm">
              <option value="Please select">Please select</option>
              <option value="Option #1">Option #1</option>
              <option value="Option #2">Option #2</option>
              <option value="Option #3">Option #3</option>
              <option value="Option #4">Option #4</option>
              <option value="Option #5">Option #5</option>
            </select>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please select one option</div>
        </fieldset>

        <!-- TEXTAREAS (SM) -->
        <div class="h2">Textarea (small)</div>
        <fieldset class="c-fieldset">
          <label class="c-label" for="messageSm">Message:</label>
          <div class="c-textarea c-textarea--sm c-textarea--invalid">
            <textarea id="messageSm"></textarea>
          </div>
          <div class="c-validation c-validation--invalid" for="postcode">Please leave a message</div>
        </fieldset>
      </div>
    </div>
  </div>
</div>
```
@adam-parsons-ca please can you confirm these are as expected based on Figma visuals.

**Comment by adam-parsons-ca** [comment]
Looks good @ChrisBrettCA . Only thing is they are missing is the icon as below.
image

I'm going to ignore the small ones for now as they aren't components yet even though the code for them exists.

Also just checking if the user for instances starts entering info or makes a selection will the validation styling be removed?

**Comment by ChrisBrettCA** [comment]
@adam-parsons-ca 
> Also just checking if the user for instances starts entering info or makes a selection will the validation styling be removed?

This will be implementation specific. I suggest however that we do come up with a standard based on what we'd like to happen and document the "behaviour" of components that often have state. This shouldn't be limited to just this component either. We'll need to think about whether this documentation is required for any other components. I will add the icon in once the PR #130 has been `merged` in as it depends on that.

**Pull request merged**